### PR TITLE
[dims] Fix large array inputs

### LINKED ISSUE #88586: functorch.dim breaks split of "big" tensors on cpu
### 🐛 Describe the bug

After importing `functorch.dim`, splitting relatively "big" tensors fails:

This works on nightly
```python
import torch
total = 0
l = []
while total < 6400:
    l.append(torch.randint(2, 10, (1,)).item())
    total += l[-1]
x = torch.randn(total, 1)
x.split(l, 0)
```

but this fails
```python
import functorch.dim
import torch
total = 0
l = []
while total < 6400:
    l.append(torch.randint(2, 10, (1,)).item())
    total += l[-1]
x = torch.randn(total, 1)
x.split(l, 0)
```

Here's the error message:
```
libc++abi: terminating with uncaught exception of type c10::Error: allocated_ <= ARENA_MAX_SIZE INTERNAL ASSERT FAILED at "/my_env/pytorch/pytorch/pytorch/functorch/csrc/dim/arena.h":227, please report a bug to PyTorch. 
Exception raised from allocate at /my_env/pytorch/pytorch/pytorch/functorch/csrc/dim/arena.h:227 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 81 (0x105555071 in libc10.dylib)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 197 (0x105553845 in libc10.dylib)
frame #2: c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, c10::detail::CompileTimeEmptyString) + 9 (0x10c89e459 in _C.cpython-38-darwin.so)
frame #3: py_split(_object*, _object* const*, long, _object*) + 3599 (0x10c8b6c7f in _C.cpython-38-darwin.so)
frame #4: cfunction_vectorcall_FASTCALL_KEYWORDS + 122 (0x1048bb79b in Python)
frame #5: method_vectorcall + 135 (0x10488f897 in Python)
frame #6: call_function + 346 (0x10492d5ec in Python)
frame #7: _PyEval_EvalFrameDefault + 30081 (0x10492a1d5 in Python)
frame #8: _PyEval_EvalCodeWithName + 1958 (0x10492e11b in Python)
frame #9: _PyFunction_Vectorcall + 228 (0x10488d94c in Python)
frame #10: call_function + 346 (0x10492d5ec in Python)
frame #11: _PyEval_EvalFrameDefault + 30273 (0x10492a295 in Python)
frame #12: _PyEval_EvalCodeWithName + 1958 (0x10492e11b in Python)
frame #13: _PyFunction_Vectorcall + 228 (0x10488d94c in Python)
frame #14: call_function + 346 (0x10492d5ec in Python)
frame #15: _PyEval_EvalFrameDefault + 30273 (0x10492a295 in Python)
frame #16: _PyEval_EvalCodeWithName + 1958 (0x10492e11b in Python)
frame #17: _PyFunction_Vectorcall + 228 (0x10488d94c in Python)
frame #18: method_vectorcall + 372 (0x10488f984 in Python)
frame #19: PyVectorcall_Call + 108 (0x10488d22d in Python)
frame #20: _PyEval_EvalFrameDefault + 31000 (0x10492a56c in Python)
frame #21: _PyEval_EvalCodeWithName + 1958 (0x10492e11b in Python)
frame #22: _PyFunction_Vectorcall + 228 (0x10488d94c in Python)
frame #23: PyVectorcall_Call + 108 (0x10488d22d in Python)
frame #24: _PyEval_EvalFrameDefault + 31000 (0x10492a56c in Python)
frame #25: _PyEval_EvalCodeWithName + 1958 (0x10492e11b in Python)
frame #26: _PyFunction_Vectorcall + 228 (0x10488d94c in Python)
frame #27: call_function + 346 (0x10492d5ec in Python)
frame #28: _PyEval_EvalFrameDefault + 30479 (0x10492a363 in Python)
frame #29: _PyEval_EvalCodeWithName + 1958 (0x10492e11b in Python)
frame #30: _PyFunction_Vectorcall + 228 (0x10488d94c in Python)
frame #31: call_function + 346 (0x10492d5ec in Python)
frame #32: _PyEval_EvalFrameDefault + 30081 (0x10492a1d5 in Python)
frame #33: function_code_fastcall + 106 (0x10488d7e6 in Python)
frame #34: method_vectorcall + 135 (0x10488f897 in Python)
frame #35: call_function + 346 (0x10492d5ec in Python)
frame #36: _PyEval_EvalFrameDefault + 30081 (0x10492a1d5 in Python)
frame #37: _PyEval_EvalCodeWithName + 1958 (0x10492e11b in Python)
frame #38: _PyFunction_Vectorcall + 228 (0x10488d94c in Python)
frame #39: _PyObject_FastCallDict + 127 (0x10488cdf3 in Python)
frame #40: _PyObject_Call_Prepend + 131 (0x10488e158 in Python)
frame #41: slot_tp_call + 71 (0x1048cfa84 in Python)
frame #42: _PyObject_MakeTpCall + 274 (0x10488cfa1 in Python)
frame #43: call_function + 804 (0x10492d7b6 in Python)
frame #44: _PyEval_EvalFrameDefault + 30479 (0x10492a363 in Python)
frame #45: function_code_fastcall + 106 (0x10488d7e6 in Python)
frame #46: call_function + 346 (0x10492d5ec in Python)
frame #47: _PyEval_EvalFrameDefault + 30053 (0x10492a1b9 in Python)
frame #48: function_code_fastcall + 106 (0x10488d7e6 in Python)
frame #49: PyVectorcall_Call + 108 (0x10488d22d in Python)
frame #50: _PyEval_EvalFrameDefault + 31000 (0x10492a56c in Python)
frame #51: _PyEval_EvalCodeWithName + 1958 (0x10492e11b in Python)
frame #52: _PyFunction_Vectorcall + 228 (0x10488d94c in Python)
frame #53: call_function + 346 (0x10492d5ec in Python)
frame #54: _PyEval_EvalFrameDefault + 30479 (0x10492a363 in Python)
frame #55: _PyEval_EvalCodeWithName + 1958 (0x10492e11b in Python)
frame #56: _PyFunction_Vectorcall + 228 (0x10488d94c in Python)
frame #57: call_function + 346 (0x10492d5ec in Python)
frame #58: _PyEval_EvalFrameDefault + 30081 (0x10492a1d5 in Python)
frame #59: function_code_fastcall + 106 (0x10488d7e6 in Python)
frame #60: method_vectorcall + 135 (0x10488f897 in Python)
frame #61: call_function + 346 (0x10492d5ec in Python)
frame #62: _PyEval_EvalFrameDefault + 30081 (0x10492a1d5 in Python)
frame #63: _PyEval_EvalCodeWithName + 1958 (0x10492e11b in Python)

Fatal Python error: Aborted

Current thread 0x0000000109708600 (most recent call first):
```

CC: @tcbegley @zou3519 @ezyang @zdevito @Chillee 

### Versions

Collecting environment information...
PyTorch version: 1.14.0.dev20221107
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6 (x86_64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.102)
CMake version: version 3.22.5
Libc version: N/A

Python version: 3.8.3 (default, May 29 2020, 14:17:09)  [Clang 11.0.3 (clang-1103.0.32.59)] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] functorch==0.3.0a0+325dcc2
[pip3] gpytorch==1.5.1
[pip3] mypy==0.931
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] pytorch-sphinx-theme==0.0.24
[pip3] torch==1.14.0.dev20221107
[pip3] torch-tb-profiler==0.2.1
[pip3] torchaudio==0.14.0.dev20221027
[pip3] torchmetrics==0.9.2
[pip3] torchrec==0.0.1a0+ed1533b
[pip3] torchrl==0.0.2a0+2f31613
[pip3] torchsnapshot==0.0.3
[pip3] torchvision==0.15.0.dev20221107
[pip3] torchx-nightly==2022.7.7
[pip3] xitorch==0.3.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.4.0           hecd8cb5_637  
[conda] mkl-service               2.4.0            py38h9ed2024_0  
[conda] mkl_fft                   1.3.1            py38h4ab4a9b_0  
[conda] mkl_random                1.2.2            py38hb2f4e1b_0  
[conda] numpy                     1.21.5           py38h2e5f0a9_2  
[conda] numpy-base                1.21.5           py38h3b1a694_2  
[conda] pytorch                   1.11.0                  py3.8_0    pytorch
[conda] torchaudio                0.11.0                 py38_cpu    pytorch
[conda] torchvision               0.12.0                 py38_cpu    pytorch

cc @zou3519 @Chillee @samdow @soumith

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88596

Variable length arguments can overflow the arena being used to keep overhead
low for torch dims. If we hit this case, we know the amount of work being done
is already relatively big, so we just fallback to standard memory allocation.

Fixes #88586

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
test failure seems real, no?

```
2022-12-06T23:05:21.2189060Z   test_big_split (__main__.TestMin) ... /var/lib/jenkins/workspace/functorch/csrc/dim/arena.h:225:27: runtime error: index 6016 out of bounds for type 'char [4096]'
2022-12-06T23:05:21.2189678Z     #0 0x7ffa9b5389ea in py::handle* Arena::allocate<py::handle>(int) (/opt/conda/lib/python3.7/site-packages/functorch/_C.cpython-37m-x86_64-linux-gnu.so+0xb69ea)
2022-12-06T23:05:21.2192621Z     #1 0x7ffa9b51d715 in Slice<py::handle>::append(Arena&, py::handle) (/opt/conda/lib/python3.7/site-packages/functorch/_C.cpython-37m-x86_64-linux-gnu.so+0x9b715)
2022-12-06T23:05:21.2193595Z     #2 0x7ffa9b50b0c5 in py_split(_object*, _object* const*, long, _object*) (/opt/conda/lib/python3.7/site-packages/functorch/_C.cpython-37m-x86_64-linux-gnu.so+0x890c5)
2022-12-06T23:05:21.2194333Z     #3 0x559b4ffb66eb in _PyMethodDef_RawFastCallKeywords /home/builder/tkoch/workspace/python_1648536129212/work/Objects/call.c:659
2022-12-06T23:05:21.2194966Z     #4 0x559b4ffec13f in _PyCFunction_FastCallKeywords /home/builder/tkoch/workspace/python_1648536129212/work/Objects/call.c:732
2022-12-06T23:05:21.2195368Z     #5 0x559b4ffec13f in _PyObject_FastCallKeywords /home/builder/tkoch/workspace/python_1648536129212/work/Objects/call.c:156
2022-12-06T23:05:21.2195730Z     #6 0x559b4ffecbd0 in call_function /home/builder/tkoch/workspace/python_1648536129212/work/Python/ceval.c:4619
2022-12-06T23:05:21.2196100Z     #7 0x559b50034701 in _PyEval_EvalFrameDefault /home/builder/tkoch/workspace/python_1648536129212/work/Python/ceval.c:3093
2022-12-06T23:05:21.2196682Z     #8 0x559b4ffa58d6 in PyEval_EvalFrameEx /home/builder/tkoch/workspace/python_1648536129212/work/Python/ceval.c:547
2022-12-06T23:05:21.2197428Z     #9 0x559b4ffa58d6 in function_code_fastcall /home/builder/tkoch/workspace/python_1648536129212/work/Objects/call.c:283
2022-12-06T23:05:21.2197844Z     #10 0x559b4ffa58d6 in _PyFunction_FastCallKeywords /home/builder/tkoch/workspace/python_1648536129212/work/Objects/call.c:408
2022-12-06T23:05:21.2198231Z     #11 0x559b50030394 in call_function /home/builder/tkoch/workspace/python_1648536129212/work/Python/ceval.c:4616
2022-12-06T23:05:21.2198732Z  
```

**Comment by zdevito** [comment]
Yeah, I wonder why it didn't show up the first time these tests ran...
